pkg/frontend/querymiddleware/shard_active_series.go

```diff
@@ -374,3 +417,14 @@ func (s *shardActiveSeriesMiddleware) writeMergedResponse(ctx context.Context, c
 
 	stream.WriteObjectEnd()
 }
+
+func (s *shardActiveSeriesMiddleware) getBufferFromPool() []byte {
+	return unsafe.Slice(s.bufferPool.Get().(*byte), jsoniterMaxBufferPoolSize)
+}
+
+func (s *shardActiveSeriesMiddleware) putBuffer(buf []byte) {
+	if cap(buf) > jsoniterMaxBufferPoolSize {
+		return
+	}
+	s.bufferPool.Put(unsafe.SliceData(buf))
+}
```

Comment:
I'm not super familiar with how sync.Pool works, is there's a reason to use unsafe here? 

Doesn't this make the slice data unreachable from gc perspective, meaning it could be collected while still referenced in the pool and cause a segfault on the next access?

Reply:
> I'm not super familiar with how sync.Pool works, is there a reason to use unsafe here?

The reason I opted to use `unsafe` is to avoid having to allocate memory when returning the buffers back to the pool, as the `sync.Put` method takes an `interface{}` as parameter, and those are escaped to heap by the compiler so that a reference can be stored into it.  

By making use of  `unsafe.SliceData`, we can directly extract the memory reference of the slice backed array, which can be wrapped by an `interface{}` without having to do any extra allocation. 

> Doesn't this make the slice data unreachable from gc perspective, meaning it could be collected while still referenced in the pool and cause a segfault on the next access?

That's a good question. As far as I can tell, this pointer is seen by the GC like any other. Or at least, there's no comment in `unsafe.SliceData` method that warns or suggests otherwise.

Reply:
Thanks for the explanation. I spent some time playing with the code and running benchmarks, and all my attempts to remove unsafe were indeed worse. I understand it's not dangerous because we're keeping a reference to a properly typed pointer, and that makes the slice's backing array visible to the gc. :+1:

Reply:
Well, finally in cdee5f64192053d1938d8b4f8c351b5f8c22081b I opted to store a reference to the slice header instead of the backed array, in order to avoid the use of `unsafe`, as it seems prone to confusion.

The only downside of this approach is that each entry in the pool requires an extra 24 bytes, but in this case, that's no a concern.

Regarding performance nothing should be affected by this change. 👍

Reply:
Ah, clever. I'm glad to not have more unsafe code 😄